### PR TITLE
bug fixes of applied jobs folder - Release branch fixes

### PR DESCRIPTION
**Pull request by surapuramakhil**
contains fixes in applied jobs folder feature

Added Job context (context object pattern) for easy access of data throughout the flow. In future we can make JobContext as context object

### CODE REVIEW

**Review by 49Simon — CHANGES_REQUESTED** [review]
There are minor changes needed related to consistency/formatting. I tested it and it works good. Really good implementation and much needed enhancement on job applications. Great work @surapuramakhil

File: src/ai_hawk/job_manager.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import json
 import os
 import random
+from sys import exc_info
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/job_manager.py:4]: done

File: src/ai_hawk/job_manager.py [hunk 2]
```diff
@@ -472,39 +473,40 @@ def next_job_page(self, position, location, job_page):
...
-            job_id = urllib.parse.parse_qs(urllib.parse.urlparse(link).query).get('currentJobId', [''])[0]
-            logger.debug(f"Job ID extracted: {job_id}")
-        except Exception as e:
-            logger.warning(f"Failed to extract job ID: {e}")
+        job = Job()
 
         try:
             logger.trace(job_tile.get_attribute('outerHTML'))
-            job_title = job_tile.find_element(By.CLASS_NAME, 'job-card-list__title').find_element(By.TAG_NAME, 'strong').text
+            job.title = job_tile.find_element(By.CLASS_NAME, 'job-card-list__title').find_element(By.TAG_NAME, 'strong').text
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/job_manager.py:483]: ah its existing code. update it anyway. we might need wrapper (repeated code) . we can add that later

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 3]
```diff
@@ -19,7 +19,9 @@
 from selenium.webdriver.support import expected_conditions as EC
 from selenium.webdriver.support.ui import Select, WebDriverWait
 
+from jobContext import JobContext
 from job_application import JobApplication
+import job_application
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:24]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 4]
```diff
@@ -704,20 +719,21 @@ def _find_and_handle_radio_question(self, section: WebElement) -> bool:
...
-            self.job_application.save_application_data({'type': 'radio', 'question': question_text, 'answer': answer})
+            job_application.save_application_data({'type': 'radio', 'question': question_text, 'answer': answer})
             self._select_radio(radios, answer)
             logger.debug("Selected new radio answer")
             return True
         return False
 
-    def _find_and_handle_textbox_question(self, section: WebElement) -> bool:
+    def _find_and_handle_textbox_question(self,job_context : JobContext, section: WebElement) -> bool:
+        job = job_context.job
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:736]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 5]
```diff
@@ -645,40 +660,39 @@ def split_text_by_width(text, font, font_size, max_width):
             logger.debug(f"Uploading cover letter from path: {file_path_pdf}")
             element.send_keys(os.path.abspath(file_path_pdf))
             job.cover_letter_path = os.path.abspath(file_path_pdf)
-            self.job_application.cover_letter_path = os.path.abspath(file_path_pdf)
+            job_context.job_application.cv_path = os.path.abspath(file_path_pdf)
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:663]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 6]
```diff
@@ -460,17 +470,21 @@ def _handle_upload_fields(self, element: WebElement, job) -> None:
...
             elif 'cover' in output:
                 logger.debug("Uploading cover letter")
-                self._create_and_upload_cover_letter(element, job)
+                self._create_and_upload_cover_letter(element, job_context)
 
         logger.debug("Finished handling upload fields")
 
-    def _create_and_upload_resume(self, element, job : Job):
+    def _create_and_upload_resume(self, element, job_context : JobContext):
+        job : Job = job_context.job
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:486]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 7]
```diff
@@ -122,18 +127,18 @@ def job_apply(self, job: Job):
...
         try:
 
             self.driver.execute_script("document.activeElement.blur();")
             logger.debug("Focus removed from the active element")
 
-            self.check_for_premium_redirect(job)
+            self.check_for_premium_redirect(job_context)
 
-            easy_apply_button = self._find_easy_apply_button(job)
+            easy_apply_button = self._find_easy_apply_button(job_context)
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:139]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 8]
```diff
@@ -439,7 +448,8 @@ def _is_upload_field(self, element: WebElement) -> bool:
         logger.debug(f"Element is upload field: {is_upload}")
         return is_upload
 
-    def _handle_upload_fields(self, element: WebElement, job) -> None:
+    def _handle_upload_fields(self, element: WebElement, job_context: JobContext) -> None:
+        job = job_context.job
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:452]: done

File: src/ai_hawk/linkedIn_easy_applier.py [hunk 9]
```diff
@@ -645,40 +660,39 @@ def split_text_by_width(text, font, font_size, max_width):
...
             logger.debug("Handled date question")
             return
-
-        if self._find_and_handle_dropdown_question(section):
+        if self._find_and_handle_dropdown_question(job_context, section):
             logger.debug("Handled dropdown question")
             return
 
-    def _handle_terms_of_service(self, element: WebElement) -> bool:
+    def _handle_terms_of_service(self,job_context: JobContext, element: WebElement) -> bool:
```
[the remark this answers is not in the document]
Comment (surapuramakhil) [reply src/ai_hawk/linkedIn_easy_applier.py:695]: its pattern / design. Now or tomorrow, we need job_context. typically job_context flows through entire control flow for providing context. I kept it for consistency. tomorrow we should be adding an anti-pattern etc.

**Pull request merged**